File: src/runtime/rs/std/math_native.rs
pub const pi: f64 = ::std::f64::consts::PI;
pub const e: f64 = ::std::f64::consts::E;

pub trait ToF64 {
    fn to_f64(self) -> f64;
}

impl ToF64 for f64 {
    fn to_f64(self) -> f64 {
        self
    }
}

impl ToF64 for f32 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for i64 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for i32 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for i16 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for i8 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for u64 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for u32 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for u16 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for u8 {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for usize {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

impl ToF64 for isize {
    fn to_f64(self) -> f64 {
        self as f64
    }
}

pub fn sin<T: ToF64>(v: T) -> f64 {
    v.to_f64().sin()
}

pub fn cos<T: ToF64>(v: T) -> f64 {
    v.to_f64().cos()
}

pub fn tan<T: ToF64>(v: T) -> f64 {
    v.to_f64().tan()
}

pub fn sqrt<T: ToF64>(v: T) -> f64 {
    v.to_f64().sqrt()
}

pub fn exp<T: ToF64>(v: T) -> f64 {
    v.to_f64().exp()
}

pub fn log<T: ToF64>(v: T) -> f64 {
    v.to_f64().ln()
}

pub fn log10<T: ToF64>(v: T) -> f64 {
    v.to_f64().log10()
}

pub fn fabs<T: ToF64>(v: T) -> f64 {
    v.to_f64().abs()
}

pub fn floor<T: ToF64>(v: T) -> f64 {
    v.to_f64().floor()
}

pub fn ceil<T: ToF64>(v: T) -> f64 {
    v.to_f64().ceil()
}

pub fn pow(a: f64, b: f64) -> f64 {
    a.powf(b)
}

File: src/runtime/go/std/math_native.go
// math_native.go: @extern delegation for pytra.std.math + pytra.std.random.
// Hand-written native implementation.
package main

import (
	"math"
	"math/rand"
	"time"
)

func py_sqrt(x any) float64  { return math.Sqrt(_toF64(x)) }
func py_sin(x any) float64   { return math.Sin(_toF64(x)) }
func py_cos(x any) float64   { return math.Cos(_toF64(x)) }
func py_tan(x any) float64   { return math.Tan(_toF64(x)) }
func py_atan2(y, x any) float64 { return math.Atan2(_toF64(y), _toF64(x)) }
func py_floor(x any) float64 { return math.Floor(_toF64(x)) }
func py_ceil(x any) float64  { return math.Ceil(_toF64(x)) }
func py_pow(x, y any) float64 { return math.Pow(_toF64(x), _toF64(y)) }
func py_exp(x any) float64   { return math.Exp(_toF64(x)) }
func py_log(x any) float64   { return math.Log(_toF64(x)) }
func py_fabs(x any) float64  { return math.Abs(_toF64(x)) }
func py_pi() float64                 { return math.Pi }

var _pytra_rng = rand.New(rand.NewSource(time.Now().UnixNano()))

func py_random() float64         { return _pytra_rng.Float64() }
func py_randint(a, b int64) int64 { return a + _pytra_rng.Int63n(b-a+1) }
func py_seed(s int64)            { _pytra_rng = rand.New(rand.NewSource(s)) }
